### PR TITLE
docs(collection): validate options

### PR DESCRIPTION
**Pull request by kvwalker**

### CODE REVIEW

File: lib/mongo_client.js [hunk 1]
```diff
@@ -391,6 +391,7 @@ MongoClient.connect = function(url, options, callback) {
  * Starts a new session on the server
  *
  * @param {SessionOptions} [options] optional settings for a driver session
+ * @param {boolean} [options.causalConsistency=true] Enables or disables causal consistency for the session.
```
Comment (daprahamian) [inline lib/mongo_client.js]: `causalConsistency` is already defined on `SessionOptions`. We do not need to duplicate it here.

File: lib/collection.js [hunk 2]
```diff
@@ -2079,6 +2102,8 @@ Collection.prototype.initializeUnorderedBulkOp = function(options) {
  *
  * @method
  * @param {object} [options] Optional settings.
+ * @param {boolean} [options.bypassDocumentValidation=false] Allow driver to bypass schema validation in MongoDB 3.2 or higher.
+ * @param {boolean} [options.checkKeys=true] Enforce document key checking.
```
Comment (daprahamian) [inline lib/collection.js]: are `checkKeys` and `bypassDocumentValidation` bson options?

Comment (kvwalker) [reply lib/collection.js]: `checkKeys` is so I'll remove that. `bypassDocumentValidation` is not.

File: lib/collection.js [hunk 3]
```diff
@@ -267,24 +267,24 @@ const DEPRECATED_FIND_OPTIONS = ['maxScan', 'fields', 'snapshot'];
  * @param {object} [options.projection] The fields to return in the query. Object of fields to include or exclude (not both), {'a':1}
  * @param {object} [options.fields] **Deprecated** Use `options.projection` instead
  * @param {number} [options.skip=0] Set to skip N documents ahead in your query (useful for pagination).
- * @param {Object} [options.hint] Tell the query to use specific indexes in the query. Object of indexes to use, {'_id':1}
- * @param {boolean} [options.explain=false] Explain the query instead of returning the data.
```
Comment (mbroadst) [inline lib/collection.js]: Do these indeed not have defaults? Is the documentation referencing a server default here?

File: lib/collection.js [hunk 4]
```diff
@@ -727,7 +727,8 @@ Collection.prototype.insert = deprecate(function(docs, options, callback) {
  * @param {number} [options.wtimeout] The write concern timeout.
  * @param {boolean} [options.j=false] Specify a journal write concern.
  * @param {boolean} [options.bypassDocumentValidation=false] Allow driver to bypass schema validation in MongoDB 3.2 or higher.
- * @param {Array} [options.arrayFilters] optional list of array filters referenced in filtered positional operators
+ * @param {array|object} [options.arrayFilters] optional list of array filters referenced in filtered positional operators
```
Comment (mbroadst) [inline lib/collection.js]: The documentation doesn't seem updated to reflect the addition of a supported `object` type

File: lib/collection.js [hunk 5]
```diff
@@ -727,7 +727,8 @@ Collection.prototype.insert = deprecate(function(docs, options, callback) {
...
+ * @param {object} [options.collation] Specify collation (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
```
Comment (mbroadst) [inline lib/collection.js]: this is an `insert` operation, not an `update` operation (doc string)

File: lib/collection.js [hunk 6]
```diff
@@ -791,7 +793,9 @@ Collection.prototype.replaceOne = function(filter, doc, options, callback) {
  * @param {(number|string)} [options.w] The write concern.
  * @param {number} [options.wtimeout] The write concern timeout.
  * @param {boolean} [options.j=false] Specify a journal write concern.
- * @param {Array} [options.arrayFilters] optional list of array filters referenced in filtered positional operators
+ * @param {array|object} [options.arrayFilters] optional list of array filters referenced in filtered positional operators
+ * @param {boolean} [options.bypassDocumentValidation=false] Allow driver to bypass schema validation in MongoDB 3.2 or higher.
+ * @param {object} [options.collation] Specify collation (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
```
Comment (mbroadst) [inline lib/collection.js]: `replace` operation, not `update` operation

File: lib/collection.js [hunk 7]
```diff
@@ -907,6 +912,7 @@ Collection.prototype.removeOne = Collection.prototype.deleteOne;
  * @param {(number|string)} [options.w] The write concern.
  * @param {number} [options.wtimeout] The write concern timeout.
  * @param {boolean} [options.j=false] Specify a journal write concern.
+ * @param {object} [options.collation] Specify collation (MongoDB 3.4 or higher) settings for update operation (see 3.4 documentation for available fields).
```
Comment (mbroadst) [inline lib/collection.js]: `remove` operation

File: lib/collection.js (hunk 3 above)
Comment (kvwalker) [reply lib/collection.js]: `explain` isn't defined as an option in the CRUD spec and is not mentioned in the [server docs](https://docs.mongodb.com/v3.4/reference/method/db.collection.find/).

File: lib/collection.js (hunk 4 above)
Comment (kvwalker) [reply lib/collection.js]: Good catch. I'll remove that.

File: lib/collection.js (hunk 5 above)
Comment (kvwalker) [reply lib/collection.js]: This is for `updateOne`.

File: lib/collection.js (hunk 6 above)
Comment (kvwalker) [reply lib/collection.js]: This is for `updateMany`.

File: lib/collection.js (hunk 3 above)
Comment (mbroadst) [reply lib/collection.js]: I guess my question is more: Do we actually set these values to a default value in the driver somewhere? If we do, and we are removing the info here, we are inconsistent. If we do set the values, and should not, we should not only remove the documentation but also the documentation

File: lib/collection.js (hunk 5 above)
Comment (mbroadst) [reply lib/collection.js]: oh whoops, github concatenated the two doc sections

File: lib/collection.js (hunk 6 above)
Comment (mbroadst) [reply lib/collection.js]: concatenated again, but this comment should be move up to the previous section where it is documented as an `update` command but is for `replaceOne`

File: lib/collection.js (hunk 3 above)
Comment (kvwalker) [reply lib/collection.js]: I don't see anywhere that we set a default for `explain` (or any of the below options for which I've removed the default).

Comment (mbroadst) [reply lib/collection.js]: :+1:

**Pull request merged**